File: plugins/moonbit-skills/scripts/check-pipeline-consistency.py
#!/usr/bin/env python3
"""
Consistency checker for MoonBit Skills repository.

Verifies:
  - All JSON files parse correctly
  - All skill frontmatter has name/description
  - Skill directory name matches frontmatter name
  - Skill count consistency across routing, orchestration, README
  - E1-E6 consistency across verify, orchestration
  - Document reference paths exist
  - Command parameters match references/commands.md definitions
  - No absolute local paths, no dead file:/// links
"""

import json
import os
import re
import sys
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parent.parent
SKILLS_DIR = REPO_ROOT / "skills"
REFERENCES_DIR = REPO_ROOT / "references"
SCHEMAS_DIR = REPO_ROOT / "schemas"


def check_json_syntax():
    """Check all JSON files parse correctly."""
    errors = []
    for path in REPO_ROOT.rglob("*.json"):
        if ".git" in path.parts:
            continue
        try:
            with open(path, "r", encoding="utf-8") as f:
                json.load(f)
        except json.JSONDecodeError as e:
            errors.append(f"Invalid JSON: {path.relative_to(REPO_ROOT)}: {e}")
    return errors


def check_skill_frontmatter():
    """Check all skill SKILL.md files have valid frontmatter."""
    errors = []
    skill_dirs = set()

    for skill_dir in sorted(SKILLS_DIR.iterdir()):
        if not skill_dir.is_dir():
            continue
        skill_md = skill_dir / "SKILL.md"
        if not skill_md.exists():
            errors.append(f"Missing SKILL.md in {skill_dir.name}")
            continue

        content = skill_md.read_text(encoding="utf-8")

        # Extract frontmatter
        m = re.match(r'^---\s*\n(.*?)\n---', content, re.DOTALL)
        if not m:
            errors.append(f"Missing or invalid frontmatter in {skill_md.relative_to(REPO_ROOT)}")
            continue

        fm = m.group(1)

        # Check name
        name_match = re.search(r'^name:\s*(\S+)', fm, re.MULTILINE)
        if not name_match:
            errors.append(f"Missing 'name' in frontmatter: {skill_md.relative_to(REPO_ROOT)}")

        # Check description
        desc_match = re.search(r'^description:\s*(.+?)$', fm, re.MULTILINE)
        if not desc_match:
            errors.append(f"Missing 'description' in frontmatter: {skill_md.relative_to(REPO_ROOT)}")

        # Check directory name matches skill name
        # Convention: directory is short name (e.g. 'code-review'), frontmatter name is 'moonbit-code-review'
        # 'using-moonbit-skills' is the full name in both
        if name_match:
            skill_name = name_match.group(1)
            dir_name = skill_dir.name
            # Expected: either exact match (using-moonbit-skills) or 'moonbit-<dir_name>' == skill_name
            expected_name = f"moonbit-{dir_name}"
            if skill_name != dir_name and skill_name != expected_name:
                errors.append(
                    f"Directory name '{dir_name}' does not match frontmatter name '{skill_name}': "
                    f"{skill_md.relative_to(REPO_ROOT)}"
                )

        skill_dirs.add(skill_dir.name)

    return errors, skill_dirs


def _without_fenced_code(content: str) -> str:
    """Remove fenced code blocks before interpreting Markdown links."""
    return re.sub(r"```.*?```", "", content, flags=re.DOTALL)


def check_skill_counts(skill_dirs):
    """Check skill count consistency across routing and orchestration."""
    errors = []
    known_skills = sorted(skill_dirs)
    known_count = len(known_skills)
    expected_core_count = known_count - (1 if "using-moonbit-skills" in known_skills else 0)
    fact_files = [
        REPO_ROOT / "README.md",
        REPO_ROOT / "assets" / "readme" / "hero.svg",
        REPO_ROOT / "assets" / "readme" / "section-skills.svg",
    ]
    for fact_file in fact_files:
        if not fact_file.exists():
            continue
        fact_content = fact_file.read_text(encoding="utf-8")
        if "13 个核心技能" in fact_content or "十三个核心技能" in fact_content:
            errors.append(
                f"Stale skill count in {fact_file.relative_to(REPO_ROOT)}: expected {expected_core_count} core skills"
            )

    # Check using-moonbit-skills SKILL.md
    bootstrap = SKILLS_DIR / "using-moonbit-skills" / "SKILL.md"
    if bootstrap.exists():
        content = bootstrap.read_text(encoding="utf-8")

        # Find the available skills table
        table_section = re.search(
            r'\|\s*Skill\s*\|\s*When to Use\s*\|\s*\n\|[-| ]+\|\n((?:\|.*\|.*\|\n)*)',
            content
        )
        if table_section:
            listed = [line.split("|")[1].strip().strip("`") for line in table_section.group(1).strip().split("\n")]
            # Filter out non-moonbit- skills
            moonbit_listed = [s for s in listed if s.startswith("moonbit-")]
            # Normalize: strip 'moonbit-' prefix for comparison with directory names
            listed_normalized = {s.replace("moonbit-", "", 1) for s in moonbit_listed}
            # Exclude bootstrap skill 'using-moonbit-skills' from comparison
            known_normalized = {s.replace("moonbit-", "", 1) if s.startswith("moonbit-") else s
                                for s in known_skills if s != "using-moonbit-skills"}
            missing_in_listed = known_normalized - listed_normalized
            extra_in_listed = listed_normalized - known_normalized
            if missing_in_listed:
                errors.append(f"Skills in directory but not in routing table: {sorted(missing_in_listed)}")
            if extra_in_listed:
                errors.append(f"Skills in routing table but no directory: {sorted(extra_in_listed)}")

    # Check orchestration.md
    orch_path = REFERENCES_DIR / "orchestration.md"
    if orch_path.exists():
        content = orch_path.read_text(encoding="utf-8")
        # Count moonbit- skills in the independent skills table
        table_section = re.search(
            r'\|\s*技能\s*\|\s*触发场景\s*\|\s*类型\s*\|\n\|[-| ]+\|\n((?:\|.*\|.*\|.*\|\n)*)',
            content
        )
        if table_section:
            listed_skills = []
            for line in table_section.group(1).strip().split("\n"):
                m = re.match(r'\|\s*`([^`]+)`', line)
                if m:
                    listed_skills.append(m.group(1))
            # Normalize: strip 'moonbit-' prefix for comparison with directory names
            listed_normalized = {s.replace("moonbit-", "", 1) for s in listed_skills if s.startswith("moonbit-")}
            # Exclude bootstrap skill 'using-moonbit-skills' from comparison
            known_normalized = {s.replace("moonbit-", "", 1) if s.startswith("moonbit-") else s
                                for s in known_skills if s != "using-moonbit-skills"}
            missing_in_orch = known_normalized - listed_normalized
            if missing_in_orch:
                errors.append(f"Skills missing in orchestration table: {sorted(missing_in_orch)}")

    return errors


def check_e6_consistency():
    """Check E1-E6 consistency across verify and orchestration."""
    errors = []
    verify_path = SKILLS_DIR / "verify" / "SKILL.md"
    orch_path = REFERENCES_DIR / "orchestration.md"

    if verify_path.exists():
        content = verify_path.read_text(encoding="utf-8")
        # Find E6 reference in execution order
        if "E6" not in content:
            errors.append("verify/SKILL.md: Missing E6 reference")

    if orch_path.exists():
        content = orch_path.read_text(encoding="utf-8")
        # Check E1-E6 in enhanced test section
        if "E1-E6" not in content:
            errors.append("orchestration.md: Missing E1-E6 reference")
        # Check E6 table row
        if "E6" not in content:
            errors.append("orchestration.md: Missing E6 table row")
        # Check panorama diagram
        if "E1-E6" not in content:
            errors.append("orchestration.md: Missing E1-E6 reference")

    return errors


def check_reference_paths():
    """Check that document reference paths exist."""
    errors = []
    doc_extensions = {".md", ".json", ".yaml", ".yml", ".sh", ".py"}

    # Check references from markdown files
    for md_file in REPO_ROOT.rglob("*.md"):
        if ".git" in md_file.parts:
            continue
        content = _without_fenced_code(md_file.read_text(encoding="utf-8"))

        # Find markdown links with relative paths
        for m in re.finditer(r'\[([^\]]+)\]\(([^)]+)\)', content):
            link = m.group(2)
            # Skip external URLs and anchors
            if link.startswith(("http://", "https://", "#", "mailto:")):
                continue
            # Skip absolute paths
            if link.startswith("/"):
                continue

            # Resolve relative to the markdown file's directory
            target = (md_file.parent / link).resolve()
            # Only check if it's a markdown or project file
            if target.suffix.lower() in doc_extensions or not target.suffix:
                if not target.exists():
                    # Allow paths with wildcards
                    if "*" not in link:
                        errors.append(
                            f"Broken reference in {md_file.relative_to(REPO_ROOT)}: "
                            f"'{link}' -> {target.relative_to(REPO_ROOT) if target.exists() else 'NOT FOUND'}"
                        )

    return errors


def check_commands_consistency():
    """Check that command references match references/commands.md if it exists."""
    errors = []
    commands_path = REFERENCES_DIR / "commands.md"

    if not commands_path.exists():
        return errors  # No commands reference to check against

    commands_content = commands_path.read_text(encoding="utf-8")
    known_commands = set()
    for m in re.finditer(r'`(moon\s+\S[^`]+)`', commands_content):
        known_commands.add(m.group(1))

    # Check other markdown files reference known commands
    for md_file in REPO_ROOT.rglob("*.md"):
        if ".git" in md_file.parts or md_file == commands_path:
            continue
        content = md_file.read_text(encoding="utf-8")
        for m in re.finditer(r'`(moon\s+\S[^`]+)`', content):
            cmd = m.group(1)
            # Skip common commands that might not be in commands.md
            if not any(cmd.startswith(prefix) for prefix in ["moon fmt", "moon check", "moon test", "moon info", "moon run", "moon doc", "moon add", "moon publish", "moon-audit"]):
                pass  # Not adding hard errors for now; this is advisory

    return errors


def check_no_absolute_paths():
    """Check for absolute local paths and dead file:/// links."""
    errors = []
    for md_file in REPO_ROOT.rglob("*.md"):
        if ".git" in md_file.parts:
            continue
        content = md_file.read_text(encoding="utf-8")

        # Check for file:/// links (excluding the allowed pattern for local file previews)
        for m in re.finditer(r'`?file:///[^`\n)]+`?', content):
            errors.append(
                f"file:/// link in {md_file.relative_to(REPO_ROOT)}: {m.group().strip()}"
            )

    return errors


def main():
    errors = []
    all_checks_passed = True

    print("=== MoonBit Skills Repository Consistency Checker ===\n")

    # 1. JSON syntax
    print("[1/6] Checking JSON syntax...")
    json_errors = check_json_syntax()
    if json_errors:
        for e in json_errors:
            print(f"  FAIL: {e}")
        errors.extend(json_errors)
    else:
        print("  PASS")

    # 2. Skill frontmatter
    print("[2/6] Checking skill frontmatter...")
    fm_errors, skill_dirs = check_skill_frontmatter()
    if fm_errors:
        for e in fm_errors:
            print(f"  FAIL: {e}")
        errors.extend(fm_errors)
    else:
        print(f"  PASS ({len(skill_dirs)} skills checked)")

    # 3. Skill count consistency
    print("[3/6] Checking skill count consistency...")
    count_errors = check_skill_counts(skill_dirs)
    if count_errors:
        for e in count_errors:
            print(f"  FAIL: {e}")
        errors.extend(count_errors)
    else:
        print("  PASS")

    # 4. E1-E6 consistency
    print("[4/6] Checking E1-E6 consistency...")
    e6_errors = check_e6_consistency()
    if e6_errors:
        for e in e6_errors:
            print(f"  FAIL: {e}")
        errors.extend(e6_errors)
    else:
        print("  PASS")

    # 5. Reference paths
    print("[5/6] Checking document reference paths...")
    ref_errors = check_reference_paths()
    if ref_errors:
        print(f"  WARN: {len(ref_errors)} issues found")
        for e in ref_errors[:10]:  # Show first 10
            print(f"    {e}")
        if len(ref_errors) > 10:
            print(f"    ... and {len(ref_errors) - 10} more")
        # Path issues are warnings, not hard failures
    else:
        print("  PASS")

    # 6. Absolute paths
    print("[6/6] Checking for absolute local paths...")
    path_errors = check_no_absolute_paths()
    if path_errors:
        for e in path_errors:
            print(f"  FAIL: {e}")
        errors.extend(path_errors)
    else:
        print("  PASS")

    print()
    if errors:
        print(f"FAILED: {len(errors)} error(s) found")
        sys.exit(1)
    else:
        print("ALL CHECKS PASSED")
        sys.exit(0)


if __name__ == "__main__":
    main()

File: plugins/moonbit-skills/scripts/check-pipeline-consistency.mjs
#!/usr/bin/env node
/*
 * Consistency checker for MoonBit Skills repository.
 *
 * Verifies:
 *   - All JSON files parse correctly
 *   - All skill frontmatter has name/description
 *   - Skill directory name matches frontmatter name
 *   - Skill count consistency across routing, orchestration, README
 *   - E1-E6 consistency across verify, orchestration
 *   - Document reference paths exist
 *   - Command parameters match references/cli/commands.md definitions
 *   - No absolute local paths, no dead file:/// links
 */

import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const REPO_ROOT = path.resolve(__dirname, "..");
const SKILLS_DIR = path.join(REPO_ROOT, "skills");
const REFERENCES_DIR = path.join(REPO_ROOT, "references");

const isDir = (p) => {
  try {
    return fs.statSync(p).isDirectory();
  } catch {
    return false;
  }
};
const isFile = (p) => {
  try {
    return fs.statSync(p).isFile();
  } catch {
    return false;
  }
};
const readText = (p) => fs.readFileSync(p, "utf-8");
const rel = (p) => path.relative(REPO_ROOT, p).split(path.sep).join("/");

/** Recursively find files matching a predicate, skipping .git and __pycache__. */
function rglob(dir, predicate) {
  const out = [];
  let entries = [];
  try {
    entries = fs.readdirSync(dir, { withFileTypes: true });
  } catch {
    return out;
  }
  for (const entry of entries) {
    if (entry.name === ".git" || entry.name === "__pycache__") continue;
    const full = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      out.push(...rglob(full, predicate));
    } else if (predicate(entry.name)) {
      out.push(full);
    }
  }
  return out;
}

const jsonFiles = () => rglob(REPO_ROOT, (n) => n.endsWith(".json"));
const mdFiles = () => rglob(REPO_ROOT, (n) => n.endsWith(".md"));

function checkJsonSyntax() {
  const errors = [];
  for (const p of jsonFiles()) {
    try {
      JSON.parse(readText(p));
    } catch (e) {
      errors.push(`${rel(p)}: ${e.message}`);
    }
  }
  return errors;
}

function checkSkillFrontmatter() {
  const errors = [];
  const skillDirs = new Set();

  let skillEntries = [];
  try {
    skillEntries = fs.readdirSync(SKILLS_DIR, { withFileTypes: true });
  } catch {
    return [errors, skillDirs];
  }

  for (const entry of skillEntries) {
    if (!entry.isDirectory()) continue;
    const skillMd = path.join(SKILLS_DIR, entry.name, "SKILL.md");
    if (!isFile(skillMd)) {
      errors.push(`Missing SKILL.md in ${entry.name}`);
      continue;
    }

    const content = readText(skillMd);

    // Extract frontmatter
    const m = /^---\s*\n([\s\S]*?)\n---/.exec(content);
    if (!m) {
      errors.push(`Missing or invalid frontmatter in ${rel(skillMd)}`);
      continue;
    }
    const fm = m[1];

    // Check name
    const nameMatch = /^name:\s*(\S+)/m.exec(fm);
    if (!nameMatch) {
      errors.push(`Missing 'name' in frontmatter: ${rel(skillMd)}`);
    }

    // Check description
    const descMatch = /^description:\s*(.+?)$/m.exec(fm);
    if (!descMatch) {
      errors.push(`Missing 'description' in frontmatter: ${rel(skillMd)}`);
    }

    // Check directory name matches skill name
    // Convention: directory is short name (e.g. 'code-review'), frontmatter name is 'moonbit-code-review'
    // 'using-moonbit-skills' is the full name in both
    if (nameMatch) {
      const skillName = nameMatch[1];
      const dirName = entry.name;
      // Expected: either exact match (using-moonbit-skills) or 'moonbit-<dir_name>' == skill_name
      const expectedName = `moonbit-${dirName}`;
      if (skillName !== dirName && skillName !== expectedName) {
        errors.push(
          `Directory name '${dirName}' does not match frontmatter name '${skillName}': ` +
            rel(skillMd)
        );
      }
    }

    skillDirs.add(entry.name);
  }

  return [errors, skillDirs];
}

/** Remove fenced code blocks before interpreting Markdown links. */
function withoutFencedCode(content) {
  return content.replace(/```[\s\S]*?```/g, "");
}

function checkSkillCounts(skillDirs) {
  const errors = [];
  const knownSkills = [...skillDirs].sort();
  const knownCount = knownSkills.length;
  const expectedCoreCount = knownCount - (knownSkills.includes("using-moonbit-skills") ? 1 : 0);

  // Check fact files for stale hard-coded counts
  const factFiles = [path.join(REPO_ROOT, "README.md")];
  for (const factFile of factFiles) {
    if (!isFile(factFile)) continue;
    const factContent = readText(factFile);
    if (factContent.includes("13 个核心技能") || factContent.includes("十三个核心技能")) {
      errors.push(
        `Stale skill count in ${rel(factFile)}: expected ${expectedCoreCount} core skills`
      );
    }
  }

  // Check using-moonbit-skills SKILL.md routing table
  const bootstrap = path.join(SKILLS_DIR, "using-moonbit-skills", "SKILL.md");
  if (isFile(bootstrap)) {
    const content = readText(bootstrap);

    // Find the available skills table
    const tableSection = /\|\s*Skill\s*\|\s*When to Use\s*\|\s*\n\|[-| ]+\|\n((?:\|.*\|.*\|\n)*)/.exec(
      content
    );
    if (tableSection) {
      const listed = tableSection[1]
        .trim()
        .split("\n")
        .map((line) => line.split("|")[1].trim().replace(/^`|`$/g, ""));
      // Filter out non-moonbit- skills
      const moonbitListed = listed.filter((s) => s.startsWith("moonbit-"));
      // Normalize: strip 'moonbit-' prefix for comparison with directory names
      const listedNormalized = new Set(moonbitListed.map((s) => s.replace(/^moonbit-/, "")));
      // Exclude bootstrap skill 'using-moonbit-skills' from comparison
      const knownNormalized = new Set(
        knownSkills
          .filter((s) => s !== "using-moonbit-skills")
          .map((s) => (s.startsWith("moonbit-") ? s.replace(/^moonbit-/, "") : s))
      );
      const missingInListed = [...knownNormalized].filter((s) => !listedNormalized.has(s)).sort();
      const extraInListed = [...listedNormalized].filter((s) => !knownNormalized.has(s)).sort();
      if (missingInListed.length) {
        errors.push(`Skills in directory but not in routing table: ${JSON.stringify(missingInListed)}`);
      }
      if (extraInListed.length) {
        errors.push(`Skills in routing table but no directory: ${JSON.stringify(extraInListed)}`);
      }
    }
  }

  // Check orchestration.md independent-skills table
  const orchPath = path.join(REFERENCES_DIR, "orchestration.md");
  if (isFile(orchPath)) {
    const content = readText(orchPath);
    const tableSection = /\|\s*技能\s*\|\s*触发场景\s*\|\s*类型\s*\|\n\|[-| ]+\|\n((?:\|.*\|.*\|.*\|\n)*)/.exec(
      content
    );
    if (tableSection) {
      const listedSkills = [];
      for (const line of tableSection[1].trim().split("\n")) {
        const m = /^\|\s*`([^`]+)`/.exec(line);
        if (m) listedSkills.push(m[1]);
      }
      // Normalize: strip 'moonbit-' prefix for comparison with directory names
      const listedNormalized = new Set(
        listedSkills
          .filter((s) => s.startsWith("moonbit-"))
          .map((s) => s.replace(/^moonbit-/, ""))
      );
      // Exclude bootstrap skill 'using-moonbit-skills' from comparison
      const knownNormalized = new Set(
        knownSkills
          .filter((s) => s !== "using-moonbit-skills")
          .map((s) => (s.startsWith("moonbit-") ? s.replace(/^moonbit-/, "") : s))
      );
      const missingInOrch = [...knownNormalized].filter((s) => !listedNormalized.has(s)).sort();
      if (missingInOrch.length) {
        errors.push(`Skills missing in orchestration table: ${JSON.stringify(missingInOrch)}`);
      }
    }
  }

  return errors;
}

function checkE6Consistency() {
  const errors = [];
  const verifyPath = path.join(SKILLS_DIR, "verify", "SKILL.md");
  const orchPath = path.join(REFERENCES_DIR, "orchestration.md");

  if (isFile(verifyPath)) {
    const content = readText(verifyPath);
    // Find E6 reference in execution order
    if (!content.includes("E6")) {
      errors.push("verify/SKILL.md: Missing E6 reference");
    }
  }

  if (isFile(orchPath)) {
    const content = readText(orchPath);
    // Check E1-E6 in enhanced test section
    if (!content.includes("E1-E6")) {
      errors.push("orchestration.md: Missing E1-E6 reference");
    }
    // Check E6 table row
    if (!content.includes("E6")) {
      errors.push("orchestration.md: Missing E6 table row");
    }
    // Check panorama diagram
    if (!content.includes("E1-E6")) {
      errors.push("orchestration.md: Missing E1-E6 reference");
    }
  }

  return errors;
}

/** Check that document reference paths exist. */
function checkReferencePaths() {
  const errors = [];
  const docExtensions = new Set([".md", ".json", ".yaml", ".yml", ".sh", ".js", ".mjs", ".ts"]);

  for (const mdFile of mdFiles()) {
    const content = withoutFencedCode(readText(mdFile));

    // Find markdown links with relative paths
    const linkRe = /\[([^\]]+)\]\(([^)]+)\)/g;
    let match;
    while ((match = linkRe.exec(content)) !== null) {
      const link = match[2];
      // Skip external URLs and anchors
      if (link.startsWith("http://") || link.startsWith("https://") || link.startsWith("#") || link.startsWith("mailto:")) {
        continue;
      }
      // Skip absolute paths
      if (link.startsWith("/")) continue;

      // Resolve relative to the markdown file's directory
      const target = path.resolve(path.dirname(mdFile), link);
      const suffix = path.extname(target).toLowerCase();
      // Only check if it's a markdown or project file
      if (docExtensions.has(suffix) || !suffix) {
        if (!isFile(target)) {
          // Allow paths with wildcards
          if (!link.includes("*")) {
            errors.push(
              `Broken reference in ${rel(mdFile)}: ` +
                `'${link}' -> ${isFile(target) ? rel(target) : "NOT FOUND"}`
            );
          }
        }
      }
    }
  }

  return errors;
}

/** Check that command references match references/cli/commands.md if it exists. */
function checkCommandsConsistency() {
  const commandsPath = path.join(REFERENCES_DIR, "cli", "commands.md");
  if (!isFile(commandsPath)) {
    return []; // No commands reference to check against
  }

  const commandsContent = readText(commandsPath);
  const knownCommands = new Set();
  const cmdRe = /`(moon\s+\S[^`]+)`/g;
  let m;
  while ((m = cmdRe.exec(commandsContent)) !== null) {
    knownCommands.add(m[1]);
  }

  // Advisory only: check other markdown files reference known commands.
  // Known command prefixes are tolerated without hard errors.
  const toleratedPrefixes = [
    "moon fmt", "moon check", "moon test", "moon info",
    "moon run", "moon doc", "moon add", "moon publish", "moon-audit",
  ];
  for (const mdFile of mdFiles()) {
    if (mdFile === commandsPath) continue;
    const content = readText(mdFile);
    const re = /`(moon\s+\S[^`]+)`/g;
    let mm;
    while ((mm = re.exec(content)) !== null) {
      const cmd = mm[1];
      if (!toleratedPrefixes.some((prefix) => cmd.startsWith(prefix))) {
        // Not adding hard errors for now; this is advisory
      }
    }
  }

  return [];
}

/** Check for absolute local paths and dead file:/// links. */
function checkNoAbsolutePaths() {
  const errors = [];
  for (const mdFile of mdFiles()) {
    const content = readText(mdFile);
    // Check for file:/// links (excluding the allowed pattern for local file previews)
    const re = /`?file:\/\/\/[^`\n)]+`?/g;
    let m;
    while ((m = re.exec(content)) !== null) {
      errors.push(`file:/// link in ${rel(mdFile)}: ${m[0].trim()}`);
    }
  }
  return errors;
}

function main() {
  const errors = [];
  const results = [];

  console.log("=== MoonBit Skills Repository Consistency Checker ===\n");

  // 1. JSON syntax
  console.log("[1/6] Checking JSON syntax...");
  const jsonErrors = checkJsonSyntax();
  if (jsonErrors.length) {
    for (const e of jsonErrors) console.log(`  FAIL: ${e}`);
    errors.push(...jsonErrors);
  } else {
    console.log("  PASS");
  }

  // 2. Skill frontmatter
  console.log("[2/6] Checking skill frontmatter...");
  const [fmErrors, skillDirs] = checkSkillFrontmatter();
  if (fmErrors.length) {
    for (const e of fmErrors) console.log(`  FAIL: ${e}`);
    errors.push(...fmErrors);
  } else {
    console.log(`  PASS (${skillDirs.size} skills checked)`);
  }

  // 3. Skill count consistency
  console.log("[3/6] Checking skill count consistency...");
  const countErrors = checkSkillCounts(skillDirs);
  if (countErrors.length) {
    for (const e of countErrors) console.log(`  FAIL: ${e}`);
    errors.push(...countErrors);
  } else {
    console.log("  PASS");
  }

  // 4. E1-E6 consistency
  console.log("[4/6] Checking E1-E6 consistency...");
  const e6Errors = checkE6Consistency();
  if (e6Errors.length) {
    for (const e of e6Errors) console.log(`  FAIL: ${e}`);
    errors.push(...e6Errors);
  } else {
    console.log("  PASS");
  }

  // 5. Reference paths
  console.log("[5/6] Checking document reference paths...");
  const refErrors = checkReferencePaths();
  if (refErrors.length) {
    console.log(`  WARN: ${refErrors.length} issues found`);
    for (const e of refErrors.slice(0, 10)) {
      console.log(`    ${e}`);
    }
    if (refErrors.length > 10) {
      console.log(`    ... and ${refErrors.length - 10} more`);
    }
    // Path issues are warnings, not hard failures
  } else {
    console.log("  PASS");
  }

  // 6. Absolute paths
  console.log("[6/6] Checking for absolute local paths...");
  const pathErrors = checkNoAbsolutePaths();
  if (pathErrors.length) {
    for (const e of pathErrors) {
      console.log(`  FAIL: ${e}`);
    }
    errors.push(...pathErrors);
  } else {
    console.log("  PASS");
  }

  console.log();
  if (errors.length) {
    console.log(`FAILED: ${errors.length} error(s) found`);
    process.exitCode = 1;
  } else {
    console.log("ALL CHECKS PASSED");
    process.exitCode = 0;
  }
  return results;
}

main();
